perf/src/packet.rs

```diff
@@ -13,21 +16,596 @@ use {
+            Self::Packet(packet) => packet.meta(),
+            Self::Bytes(packet) => packet.meta(),
+        }
+    }
+
+    #[inline]
+    pub fn size(&self) -> usize {
+        match self {
+            Self::Packet(packet) => packet.meta().size,
+            Self::Bytes(packet) => packet.size(),
+        }
+    }
+
+    pub fn deserialize_slice<T, I>(&self, index: I) -> bincode::Result<T>
+    where
+        T: serde::de::DeserializeOwned,
+        I: SliceIndex<[u8], Output = [u8]>,
+    {
+        match self {
+            Self::Packet(packet) => packet.deserialize_slice(index),
+            Self::Bytes(packet) => packet.deserialize_slice(index),
+        }
+    }
+
+    pub fn to_packet(&self) -> Packet {
+        let size = self.size();
+        let mut packet = Packet::default();
+        if let Some(data) = self.data(..) {
+            packet.buffer_mut()[..size].copy_from_slice(data);
+        }
+        *packet.meta_mut() = self.meta().clone();
+        packet.meta_mut().size = size;
+        packet
+    }
+}
+
+#[derive(Debug, Eq)]
+pub enum PacketRefMut<'a> {
+    Packet(&'a mut Packet),
+    Bytes(&'a mut BytesPacket),
+}
+
+impl<'a> PartialEq for PacketRefMut<'a> {
+    fn eq(&self, other: &PacketRefMut<'a>) -> bool {
+        self.data(..).eq(&other.data(..)) && self.meta().eq(other.meta())
+    }
+}
+
+impl<'a> From<&'a mut Packet> for PacketRefMut<'a> {
+    fn from(packet: &'a mut Packet) -> Self {
+        Self::Packet(packet)
+    }
+}
+
+impl<'a> From<&'a mut BytesPacket> for PacketRefMut<'a> {
+    fn from(packet: &'a mut BytesPacket) -> Self {
+        Self::Bytes(packet)
+    }
+}
+
+impl PacketRefMut<'_> {
+    pub fn data<I>(&self, index: I) -> Option<&<I as SliceIndex<[u8]>>::Output>
+    where
+        I: SliceIndex<[u8]>,
+    {
+        match self {
+            Self::Packet(packet) => packet.data(index),
+            Self::Bytes(packet) => packet.data(index),
+        }
+    }
+
+    #[inline]
+    pub fn meta(&self) -> &Meta {
+        match self {
+            Self::Packet(packet) => packet.meta(),
+            Self::Bytes(packet) => packet.meta(),
+        }
+    }
+
+    #[inline]
+    pub fn meta_mut(&mut self) -> &mut Meta {
+        match self {
+            Self::Packet(packet) => packet.meta_mut(),
+            Self::Bytes(packet) => packet.meta_mut(),
+        }
+    }
+
+    #[inline]
+    pub fn size(&self) -> usize {
+        match self {
+            Self::Packet(packet) => packet.meta().size,
+            Self::Bytes(packet) => packet.size(),
+        }
+    }
+
+    pub fn deserialize_slice<T, I>(&self, index: I) -> bincode::Result<T>
+    where
+        T: serde::de::DeserializeOwned,
+        I: SliceIndex<[u8], Output = [u8]>,
+    {
+        match self {
+            Self::Packet(packet) => packet.deserialize_slice(index),
+            Self::Bytes(packet) => packet.deserialize_slice(index),
+        }
+    }
+
+    #[inline]
+    pub fn copy_from_packet(&mut self, src_packet: PacketRef<'_>) {
```

Comment:
pretty sure this function shouldn't exist. The only caller is shrink_batches.
The logic is different for Packet (requires copy) and BytesPacket (can just move
Bytes). Just inline the logic there?

Reply:
Fair, I can do that.

Reply:
Removed :axe: